src/utils/bitfield.rs

```diff
@@ -0,0 +1,231 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `range` - The range of bits, starting from 0 for the rightmost one.
+    /// * `value` - An integer representing the desired state of the bits.
+    ///
+    /// # Panics
+    ///
+    /// This method will panics if the given range is invalid.
+    /// Value must also not be greater in terms of bits than the given range.
+    fn set_bits<T: RangeBounds<usize>>(self, range: T, value: Self) -> Self;
+}
+
+macro_rules! impl_bitfields {
+    (for $($t:ty),+) => {
+        $(impl BitField for $t {
+            const TYPE_SIZE: usize = core::mem::size_of::<Self>() * 8;
+
+            fn get_bit(self, idx: usize) -> bool {
+                if idx >= Self::TYPE_SIZE {
+                    panic!("Index out of range for bit fields");
+                }
+
+                (self & (1 << idx)) != 0
+            }
+
+            fn get_bits<T: RangeBounds<usize>>(self, range: T) -> Self {
+                let range = to_regular_range(&range, Self::TYPE_SIZE);
+
+                if range.start >= Self::TYPE_SIZE || range.end > Self::TYPE_SIZE {
+                    panic!("Range out of range for bit fields");
+                }
+                if range.start >= range.end {
+                    panic!("End bound should be greater than lower one for bit fields.");
+                }
+
+                // Shift away high bits then lower ones
+                let val = self << (Self::TYPE_SIZE - range.end);
+                val >> ((Self::TYPE_SIZE - range.end) + range.start)
+            }
+
+            fn set_bit(self, idx: usize, value: bool) -> Self {
+                if idx >= Self::TYPE_SIZE {
+                    panic!("Index out of range for bit fields");
+                }
+
+                if value { self | (1 << idx) } else { self & (!(1 << idx)) }
+            }
+
+            fn set_bits<T: RangeBounds<usize>>(self, range: T, value: Self) -> Self {
+                let range = to_regular_range(&range, Self::TYPE_SIZE);
+
+                if range.start >= Self::TYPE_SIZE || range.end > Self::TYPE_SIZE {
+                    panic!("Range out of range for bit fields");
+                }
+                if range.start >= range.end {
+                    panic!("End bound should be greater than lower one for bit fields.");
+                }
+                if (range.end - range.start) < Self::TYPE_SIZE && (value >> (range.end - range.start)) > 0 {
+                    panic!("Value bigger than range.");
+                }
+
+                let mask: Self = !(!0 << (Self::TYPE_SIZE - range.end) >>
+                                    (Self::TYPE_SIZE - range.end) >>
+                                    range.start << range.start);
+                (self & mask) | (value << range.start)
+            }
+        })*
+    }
+}
+
+impl_bitfields!(for u8, u16, u32, u64);
+
+fn to_regular_range<T: RangeBounds<usize>>(range: &T, maximun: usize) -> Range<usize> {
```

Comment:
Wouldn't we implement `From<RangeBounds<usize>> for Range<usize>` ?

Reply:
After some testing, whilst it is indeed possible I'm doubting about the conveyed message when implementing the `From` trait since it could be reused in other context where this particular transformation does not apply.